File: two_pointers_problems/3sum/solution.py
'''
this problem can be solved by brute force approach---O(n3)
'''


def threeSum(nums):
    l=[]
    l1=[]
    i=0
    nums.sort()
    while i<len(nums):
        j=i+1
        while j<len(nums):
            k=j+1
            while k<len(nums):
                if nums[i]+nums[j]+nums[k]==0:
                    s=str(nums[i])+" "+str(nums[j])+" "+str(nums[k])
                    if s not in l1:
                        l1.append(s)
                        l.append([nums[i],nums[j],nums[k]])
                k+=1
            j+=1
        i+=1
    return l



'''
this problem can be solved by hashmap approach so that we can 
reduce searching time ---O(n2)
space---O(n)

'''

def threeSum(nums):
    l=[]
    nums.sort()
    st=set()
    i=0
    while i<len(nums):
        j=i+1
        d={}
        while j<len(nums):
            if -nums[i]-nums[j] in d.keys():
                s=str(nums[i])+" "+str(nums[j])+" "+str(-nums[i]-nums[j])
                if s not in st:
                    st.add(s)
                    l.append([nums[i],nums[j],-nums[i]-nums[j]])
            d[nums[j]]=j
            
            j+=1
        i+=1
    return l


'''
this problem can be solved by two pointers approach---O(n2)
space--O(1)

'''



def threeSum(nums):
    i=0
    nums.sort()
    l=[]
    while i<len(nums):
        while i>0 and i<len(nums) and nums[i]==nums[i-1]:
            i+=1

        j=i+1
        k=len(nums)-1
        while j<k:
            if nums[i]+nums[j]+nums[k]==0:
                l.append([nums[i],nums[j],nums[k]])
                j+=1
                k-=1

                while j<k and nums[j]==nums[j-1]:
                    j+=1
                
                while j<k and nums[k]==nums[k+1]:
                    k-=1
               
            elif nums[i]+nums[j]+nums[k]>0:
                k-=1
            else:
                j+=1
            
        i+=1
    return l

File: two_pointers_problems/3sum/solution.js
/**
 * this problem can be solved by brute force approach
 * time complexity-----O(n3+nlogn)
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  var res = [];
  nums.sort((a, b) => a - b);
  var i = 0;
  while (i < nums.length) {
    if (i > 0 && nums[i] == nums[i - 1]) {
      i += 1;
      continue;
    }
    var j = i + 1;
    while (j < nums.length) {
      if (j - i > 1 && nums[j] == nums[j - 1]) {
        j += 1;
        continue;
      }
      var k = j + 1;
      while (k < nums.length) {
        if (k - j > 1 && nums[k] == nums[k - 1]) {
          k += 1;
          continue;
        }
        if (nums[i] + nums[j] + nums[k] == 0) {
          res.push([nums[i], nums[j], nums[k]]);
        }
        k += 1;
      }

      j += 1;
    }

    i += 1;
  }

  return res;
};

/**
 * this problem can be solved by  hashset---O(nlogn+n2*hashsearch+no_of_triplets*3)
 * space complexity--O(n+no_of_triplets*3)
 * @param {number[]} nums
 * @return {number[][]}
 */

var threeSum = function (nums) {
  nums.sort((a, b) => a - b);
  var i = 0;
  var res = new Set();

  while (i < nums.length) {
    var j = i + 1;
    var hs = new Set();
    while (j < nums.length) {
      if (hs.has(-nums[i] - nums[j])) {
        res.add(nums[i] + " " + nums[j] + " " + (-nums[i] - nums[j]));
      }
      hs.add(nums[j]);

      j += 1;
    }
    i += 1;
  }
  console.log(res);
  var final = [...res].map((str) => {
    return str.split(" ").map((num) => +num);
  });
  return final;
};

/**
 * this problem can be solved by two pointers approach--O(nlogn+n2)
 * @param {number[]} nums
 * @return {number[][]}
 */

var threeSum = function (nums) {
  nums.sort((a, b) => a - b);
  var res = [];
  var i = 0;
  while (i < nums.length) {
    if (i > 0 && nums[i] == nums[i - 1]) {
      i += 1;
      continue;
    }
    var low = i + 1;
    var high = nums.length - 1;
    while (low < high) {
      if (low - i > 1 && nums[low] == nums[low - 1]) {
        low += 1;
        continue;
      }
      if (high < nums.length - 1 && nums[high] == nums[high + 1]) {
        high -= 1;
        continue;
      }
      var sum = nums[i] + nums[low] + nums[high];
      if (sum == 0) {
        res.push([nums[i], nums[low], nums[high]]);
        low += 1;
        high -= 1;
      } else if (sum > 0) {
        high -= 1;
      } else {
        low += 1;
      }
    }
    i += 1;
  }
  return res;
};
